test/distributed/algorithms/test_join.py

```diff
@@ -0,0 +1,420 @@
+        OFFSET = 1
+
+        allreducer = AllReducer(self.device, self.process_group)
+        inputs = self.construct_uneven_inputs(BASE_NUM_INPUTS, OFFSET)
+        allreduce_total = 0
+        with _Join([allreducer]):
+            for _ in inputs:
+                allreduce_total += allreducer()
+
+        # Non-joined processes all-reduce a 1, so this rank's all-reduce total
+        # should be precisely equal to the total number of inputs processed
+        # before it joined
+        expected_total = self.world_size * BASE_NUM_INPUTS
+        # Rank i runs for i additional iterations
+        for num_joined in range(1, self.rank + 1):
+            expected_total += (self.world_size - num_joined) * OFFSET
+        self.assertEqual(allreduce_total, expected_total)
+
+    @require_n_gpus_for_nccl_backend(
+        WORLD_SIZE, BACKEND
+    )
+    def test_single_joinable_post_hooks(self):
+        r"""Tests the post-hooks of a single :class:`_Joinable`."""
+        self.dist_init(self.rank, self.world_size)
+        BASE_NUM_INPUTS = 3
+        OFFSET = 1
+
+        allreducer = AllReducer(self.device, self.process_group)
+        self.assertEqual(allreducer.post_hook_tensor.item(), BEFORE_CONSTANT)
+
+        inputs = self.construct_uneven_inputs(BASE_NUM_INPUTS, OFFSET)
+        with _Join([allreducer], run_post_hooks=True):
+            for _ in inputs:
+                allreducer()
+
+        # All processes should receive the updated `post_hook_tensor` from the
+        # last-joined process
+        self.assertEqual(allreducer.post_hook_tensor.item(), AFTER_CONSTANT)
+
+    @require_n_gpus_for_nccl_backend(
+        WORLD_SIZE, BACKEND
+    )
+    def test_single_joinable(self):
+        r"""
+        Tests the main hooks and post-hooks of a single :class:`_Joinable`
+        together.
+
+        This combines ``test_single_joinable_main_hooks()`` and
+        ``test_single_joinable_post_hooks()`` into a single test to ensure that
+        main hooks and post-hooks operate correctly together.
+        """
+        self.dist_init(self.rank, self.world_size)
+        BASE_NUM_INPUTS = 3
+        OFFSET = 1
+
+        allreducer = AllReducer(self.device, self.process_group)
+        self.assertEqual(allreducer.post_hook_tensor.item(), BEFORE_CONSTANT)
+
+        inputs = self.construct_uneven_inputs(BASE_NUM_INPUTS, OFFSET)
+        allreduce_total = 0
+        with _Join([allreducer], run_post_hooks=True):
+            for _ in inputs:
+                allreduce_total += allreducer()
+
+        expected_total = self.world_size * BASE_NUM_INPUTS
+        for num_joined in range(1, self.rank + 1):
+            expected_total += (self.world_size - num_joined) * OFFSET
+        self.assertEqual(allreduce_total, expected_total)
+        self.assertEqual(allreducer.post_hook_tensor.item(), AFTER_CONSTANT)
+
+    @require_n_gpus_for_nccl_backend(
+        WORLD_SIZE, BACKEND
+    )
+    def test_multiple_joinables(self):
```

Comment:
looks like this can also dedup with `test_single_joinable_main_hooks` by providing a list of hooks and expected results to a common function?